File: packages/snow_draw_rust_engine/src/draw/utils/list_equality.rs
#![allow(dead_code)]

use crate::draw::types::draw_point::DrawPoint;
use crate::draw::types::edit_transform::ElbowFixedSegment;
use std::collections::HashMap;
use std::hash::{BuildHasher, Hash, Hasher};

/// Returns `true` when two slices are equal element-by-element using
/// `element_equals`.
pub fn list_equals_by<T, F>(a: &[T], b: &[T], mut element_equals: F) -> bool
where
    F: FnMut(&T, &T) -> bool,
{
    if std::ptr::eq(a, b) {
        return true;
    }

    if a.len() != b.len() {
        return false;
    }

    a.iter()
        .zip(b.iter())
        .all(|(left, right)| element_equals(left, right))
}

/// Returns `true` when two slices are equal using `PartialEq`.
pub fn list_equals<T>(a: &[T], b: &[T]) -> bool
where
    T: PartialEq,
{
    list_equals_by(a, b, |left, right| left == right)
}

/// Computes a positional hash for a slice.
pub fn list_hash<T>(values: &[T]) -> u64
where
    T: Hash,
{
    hash_all(values.iter().map(hash_value))
}

/// Nullable variant of [`list_equals_by`].
pub fn nullable_list_equals_by<T, F>(a: Option<&[T]>, b: Option<&[T]>, element_equals: F) -> bool
where
    F: FnMut(&T, &T) -> bool,
{
    match (a, b) {
        (None, None) => true,
        (Some(left), Some(right)) => list_equals_by(left, right, element_equals),
        _ => false,
    }
}

/// Nullable variant of [`list_equals`].
pub fn nullable_list_equals<T>(a: Option<&[T]>, b: Option<&[T]>) -> bool
where
    T: PartialEq,
{
    nullable_list_equals_by(a, b, |left, right| left == right)
}

/// Computes a deterministic hash for a map by sorting keys by their string
/// representation before hashing entries.
pub fn map_hash<K, V, S>(map: &HashMap<K, V, S>) -> u64
where
    K: Eq + Hash + ToString,
    V: Hash,
    S: BuildHasher,
{
    if map.is_empty() {
        return 0;
    }

    let mut sorted_entries: Vec<(&K, &V)> = map.iter().collect();
    sorted_entries
        .sort_by(|(left_key, _), (right_key, _)| left_key.to_string().cmp(&right_key.to_string()));

    hash_all(sorted_entries.into_iter().map(|(key, value)| {
        let mut hasher = std::collections::hash_map::DefaultHasher::new();
        key.hash(&mut hasher);
        value.hash(&mut hasher);
        hasher.finish()
    }))
}

/// Returns `true` when two maps are equal using `value_equals` for values.
pub fn map_equals_by<K, V, SA, SB, F>(
    a: &HashMap<K, V, SA>,
    b: &HashMap<K, V, SB>,
    mut value_equals: F,
) -> bool
where
    K: Eq + Hash,
    SA: BuildHasher,
    SB: BuildHasher,
    F: FnMut(&V, &V) -> bool,
{
    if a.len() != b.len() {
        return false;
    }

    for (key, left_value) in a {
        let Some(right_value) = b.get(key) else {
            return false;
        };

        if !value_equals(left_value, right_value) {
            return false;
        }
    }

    true
}

/// Returns `true` when two maps are equal using value `PartialEq`.
pub fn map_equals<K, V, SA, SB>(a: &HashMap<K, V, SA>, b: &HashMap<K, V, SB>) -> bool
where
    K: Eq + Hash,
    V: PartialEq,
    SA: BuildHasher,
    SB: BuildHasher,
{
    map_equals_by(a, b, |left, right| left == right)
}

/// Returns `true` when two `DrawPoint` lists are equal.
pub fn point_list_equals(a: &[DrawPoint], b: &[DrawPoint]) -> bool {
    list_equals(a, b)
}

/// Returns `true` when two optional fixed-segment lists are equal.
pub fn fixed_segment_list_equals(
    a: Option<&[ElbowFixedSegment]>,
    b: Option<&[ElbowFixedSegment]>,
) -> bool {
    nullable_list_equals(a, b)
}

/// Returns whether the segment formed by `a` and `b` is horizontal.
pub fn segment_is_horizontal(a: DrawPoint, b: DrawPoint) -> bool {
    (a.y - b.y).abs() <= (a.x - b.x).abs()
}

/// Returns structural equality for optional fixed-segment lists.
///
/// The current Rust `ElbowFixedSegment` type is still a placeholder in this
/// crate, so this falls back to strict item equality for now.
pub fn fixed_segment_structure_equals(
    a: Option<&[ElbowFixedSegment]>,
    b: Option<&[ElbowFixedSegment]>,
) -> bool {
    nullable_list_equals_by(a, b, |left, right| {
        fixed_segment_structure_item_equals(left, right, None)
    })
}

/// Returns structural equality for optional fixed-segment lists with tolerance.
///
/// The current Rust `ElbowFixedSegment` type is still a placeholder in this
/// crate, so this falls back to strict item equality for now.
pub fn fixed_segment_structure_equals_with_tolerance(
    a: Option<&[ElbowFixedSegment]>,
    b: Option<&[ElbowFixedSegment]>,
    tolerance: f64,
) -> bool {
    assert!(tolerance >= 0.0, "tolerance must be non-negative.");

    nullable_list_equals_by(a, b, |left, right| {
        fixed_segment_structure_item_equals(left, right, Some(tolerance))
    })
}

fn fixed_segment_structure_item_equals(
    a: &ElbowFixedSegment,
    b: &ElbowFixedSegment,
    _tolerance: Option<f64>,
) -> bool {
    a == b
}

fn hash_value<T>(value: &T) -> u64
where
    T: Hash,
{
    let mut hasher = std::collections::hash_map::DefaultHasher::new();
    value.hash(&mut hasher);
    hasher.finish()
}

fn hash_all(values: impl IntoIterator<Item = u64>) -> u64 {
    values
        .into_iter()
        .fold(0_u64, |acc, value| acc.wrapping_mul(31).wrapping_add(value))
}

File: packages/snow_draw_engine/lib/draw/utils/list_equality.dart
import '../elements/types/arrow/elbow/elbow_fixed_segment.dart';
import '../types/draw_point.dart';

/// Shared list-equality helpers used across element data classes and
/// edit operations.
///
/// These replace the many private `_pointsEqual` / `_fixedSegmentsEqual`
/// copies that were scattered throughout the codebase.

/// Element-wise equality for lists using [elementEquals].
bool listEqualsBy<T>(
  List<T> a,
  List<T> b,
  bool Function(T left, T right) elementEquals,
) {
  if (identical(a, b)) {
    return true;
  }
  if (a.length != b.length) {
    return false;
  }
  for (var i = 0; i < a.length; i++) {
    if (!elementEquals(a[i], b[i])) {
      return false;
    }
  }
  return true;
}

/// Element-wise equality for lists using `==` comparison.
bool listEquals<T>(List<T> a, List<T> b) =>
    listEqualsBy(a, b, (left, right) => left == right);

/// Hash for a list using positional `Object.hashAll` semantics.
int listHash<T>(List<T> values) => Object.hashAll(values);

/// Nullable variant of [listEqualsBy].
bool nullableListEqualsBy<T>(
  List<T>? a,
  List<T>? b,
  bool Function(T left, T right) elementEquals,
) {
  if (a == null || b == null) {
    return a == b;
  }
  return listEqualsBy(a, b, elementEquals);
}

/// Nullable variant of [listEquals].
bool nullableListEquals<T>(List<T>? a, List<T>? b) =>
    nullableListEqualsBy(a, b, (left, right) => left == right);

/// Hash for a map using deterministic key order.
///
/// Keys are sorted by `toString()` to keep hash generation stable across runs.
int mapHash<K, V>(Map<K, V> map) {
  if (map.isEmpty) {
    return 0;
  }
  final sortedEntries = map.entries.toList(
    growable: false,
  )..sort((left, right) => left.key.toString().compareTo(right.key.toString()));
  return Object.hashAll(
    sortedEntries.map((entry) => Object.hash(entry.key, entry.value)),
  );
}

/// Element-wise equality for maps using key/value equality.
bool mapEqualsBy<K, V>(
  Map<K, V> a,
  Map<K, V> b,
  bool Function(V left, V right) valueEquals,
) {
  if (identical(a, b)) {
    return true;
  }
  if (a.length != b.length) {
    return false;
  }
  for (final entry in a.entries) {
    if (!b.containsKey(entry.key)) {
      return false;
    }
    if (!valueEquals(entry.value, b[entry.key] as V)) {
      return false;
    }
  }
  return true;
}

/// Element-wise equality for maps using `==` for values.
bool mapEquals<K, V>(Map<K, V> a, Map<K, V> b) =>
    mapEqualsBy(a, b, (left, right) => left == right);

/// Element-wise equality for [DrawPoint] lists.
bool pointListEquals(List<DrawPoint> a, List<DrawPoint> b) => listEquals(a, b);

/// Element-wise equality for nullable [ElbowFixedSegment] lists.
///
/// Uses strict `!=` comparison (index + start + end), matching the
/// semantics used by data-class `==` operators.
bool fixedSegmentListEquals(
  List<ElbowFixedSegment>? a,
  List<ElbowFixedSegment>? b,
) => nullableListEquals(a, b);

/// Whether segment [a] is horizontal based on its endpoints.
bool segmentIsHorizontal(DrawPoint a, DrawPoint b) =>
    (a.y - b.y).abs() <= (a.x - b.x).abs();

/// Structural equality for [ElbowFixedSegment] lists used by
/// arrow point transforms - compares index and axis orientation
/// rather than exact start/end positions.
bool fixedSegmentStructureEquals(
  List<ElbowFixedSegment>? a,
  List<ElbowFixedSegment>? b,
) => nullableListEqualsBy(a, b, _fixedSegmentStructureItemEquals);

/// Structural equality with axis-value tolerance, used by the arrow
/// point operation to detect meaningful segment changes.
bool fixedSegmentStructureEqualsWithTolerance(
  List<ElbowFixedSegment>? a,
  List<ElbowFixedSegment>? b, {
  double tolerance = 1.0,
}) {
  assert(tolerance >= 0, 'tolerance must be non-negative.');
  return nullableListEqualsBy(
    a,
    b,
    (left, right) =>
        _fixedSegmentStructureItemEquals(left, right, tolerance: tolerance),
  );
}

bool _fixedSegmentStructureItemEquals(
  ElbowFixedSegment a,
  ElbowFixedSegment b, {
  double? tolerance,
}) {
  if (a.index != b.index) {
    return false;
  }

  final isHorizontal = segmentIsHorizontal(a.start, a.end);
  if (isHorizontal != segmentIsHorizontal(b.start, b.end)) {
    return false;
  }

  if (tolerance == null) {
    return true;
  }

  return (_segmentAxis(a, isHorizontal) - _segmentAxis(b, isHorizontal))
          .abs() <=
      tolerance;
}

double _segmentAxis(ElbowFixedSegment segment, bool isHorizontal) {
  if (isHorizontal) {
    return (segment.start.y + segment.end.y) / 2;
  }
  return (segment.start.x + segment.end.x) / 2;
}
